pkg/meta/redis.go

```diff
@@ -2809,53 +2809,105 @@ func (m *redisMeta) ListSlices(ctx Context, slices map[Ino][]Slice, delete bool,
 		return errno(err)
 	}
 
-	return errno(m.scanDeletedSlices(ctx, func(s Slice) error {
-		slices[1] = append(slices[1], s)
+	return errno(m.scanDeletedSlices(ctx, func(ss []Slice, _ int64) (bool, error) {
+		slices[1] = append(slices[1], ss...)
 		if showProgress != nil {
-			showProgress()
+			for range ss {
+				showProgress()
+			}
 		}
-		return nil
+		return false, nil
 	}))
 }
 
-func (m *redisMeta) scanDeletedSlices(ctx context.Context, visitor func(s Slice) error) error {
-	if visitor == nil {
+func (m *redisMeta) scanDeletedSlices(ctx Context, scan deletedSliceScan) error {
+	if scan == nil {
 		return nil
 	}
-	visited := make(map[uint64]bool)
-	return m.hscan(ctx, m.delSlices(), func(keys []string) error {
-		var ss []Slice
-		for i := 0; i < len(keys); i += 2 {
-			ss = ss[:0]
-			m.decodeDelayedSlices([]byte(keys[i+1]), &ss)
-			for _, s := range ss {
-				if s.Id > 0 && !visited[s.Id] {
-					visited[s.Id] = true
-					if err := visitor(s); err != nil {
-						return err
+
+	delKeys := make(chan string, 1000)
+	c, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func() {
+		_ = m.hscan(c, m.delSlices(), func(keys []string) error {
+			for i := 0; i < len(keys); i += 2 {
+				delKeys <- keys[i]
+			}
+			return nil
+		})
+		close(delKeys)
+	}()
+
+	var ss []Slice
+	var rs []*redis.IntCmd
+	for key := range delKeys {
+		ss = ss[:0]
```

Comment:
this should be within the task func because task may run several times (transaction retry)

Reply:
Fixed